e2e/nomostest/registry_server.go

```diff
@@ -0,0 +1,358 @@
+		}
+	}
+
+	return nt.Watcher.WatchForCurrentStatus(kinds.Deployment(), TestRegistryServer, TestRegistryNamespace)
+}
+
+// PushOCIImage uses the current file system state of the provided Repository
+// to build an OCI image and push it to the current RegistryProvider. The resulting
+// OCIImage object can be used to set the spec.oci.image field on the RSync.
+func (nt *NT) PushOCIImage(repository *gitproviders.Repository) (*registryproviders.OCIImage, error) {
+	// Construct artifactDir using TmpDir. TmpDir is scoped to each test case and
+	// cleaned up after the test.
+	artifactDir := filepath.Join(nt.TmpDir, "artifacts", "oci", repository.Name)
+	if err := os.MkdirAll(artifactDir, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("creating artifact dir: %w", err)
+	}
+	image, err := registryproviders.NewImage(artifactDir, nt.Shell, repository, nt.RegistryProvider)
+	if err != nil {
+		return nil, err
+	}
+	// Track image for cleanup and for recovering from a LocalProvider crash.
+	nt.ociImages = append(nt.ociImages, image)
+
+	address, err := nt.RegistryProvider.OCIPushURL(repository.Name)
+	if err != nil {
+		return nil, fmt.Errorf("getting OCIPushURL: %w", err)
+	}
+
+	if err := image.Push(address); err != nil {
+		return nil, err
+	}
+
+	return image, nil
+}
+
+func testRegistryServerSelector() map[string]string {
+	return map[string]string{"app": TestRegistryServer}
+}
+
+func registryServer() []client.Object {
+	objs := []client.Object{
+		registryService(),
+		registryServiceAuthenticated(),
+		nginxConfigMap(),
+		registryDeployment(),
+	}
+	return objs
+}
+
+// This service exposes a port which maps to the nginx endpoint which uses HTTPS
+// but does not require authentication
+func registryService() *v1.Service {
+	service := fake.ServiceObject(
+		core.Name(TestRegistryServer),
+		core.Namespace(TestRegistryNamespace),
+	)
+	service.Spec.Selector = testRegistryServerSelector()
+	service.Spec.Ports = []v1.ServicePort{
+		{Name: "https", Port: 443, TargetPort: intstr.FromInt(RegistryHTTPSPort)},
+	}
+	return service
+}
+
+// This service exposes a port which maps to the nginx endpoint which requires
+// authentication (using contrived username/password)
+func registryServiceAuthenticated() *v1.Service {
+	service := fake.ServiceObject(
+		core.Name(TestRegistryServerAuthenticated),
+		core.Namespace(TestRegistryNamespace),
+	)
+	service.Spec.Selector = testRegistryServerSelector()
+	service.Spec.Ports = []v1.ServicePort{
+		{Name: "https", Port: 443, TargetPort: intstr.FromInt(RegistryHTTPSAuthPort)},
+	}
+	return service
+}
+
+func nginxConfigMap() *v1.ConfigMap {
+	configMap := fake.ConfigMapObject(core.Name(nginxConfigMapName),
+		core.Namespace(TestRegistryNamespace),
+		core.Labels(testRegistryServerSelector()),
+	)
+	configMap.Data = map[string]string{
+		nginxConf: nginxConfig,
+	}
+	return configMap
+}
+
+func registryDeployment() *appsv1.Deployment {
```

Comment:
Can this be a testdata artifact?

Reply:
It _can_ be, but I prefer it like this personally. With it inline it's easier to discover from the code and reuse constants IMO

Reply:
If you need to modify values you can always parse the file into a client.Object and override specific files with constants.

Reply:
Ack, I still prefer this approach